sycl/include/sycl/types.hpp

```diff
@@ -495,7 +533,12 @@ template <typename Type, int NumElements> class vec {
   template <size_t... Is>
   constexpr vec(const std::array<vec_data_t<DataT>, NumElements> &Arr,
                 std::index_sequence<Is...>)
-      : m_Data{vec_data_t<DataT>(static_cast<DataT>(Arr[Is]))...} {}
+      : m_Data{([&](vec_data_t<DataT> v) constexpr {
+          if constexpr (std::is_same_v<sycl::ext::oneapi::bfloat16, DataT>)
+            return v.value;
+          else
```

Comment:
Why do we need this? Don't we provide conversion operators that would work with the `static_cast`?

Reply:
The storage type for bfloat16 is `uint16_t` and that's the data behind the `m_Data` but the `DataT` is bfloat16. There is, intentionally, no direct conversion between those types. If you cast, the compiler will insert a call to the bfloat16 float operator() for the conversion, which is unnecessary and, in many cases, will mess up the value.  So there a couple of places the vec<bfloat16> needs to be careful.